### ISSUE TITLE
sqlite error opening connection on raspbian

### ISSUE THREAD

**Issue by yonisolo** [gnosygnu/xowa#27]
Hi,
I have a new problem,
Xowa 1.9.4 working fine on debian wheezy/jessie with standard x64 architecture but when i tried to launch it on raspbian wheezy (debian on raspberry) the launcher crached with this stack trace:
```
app init failed: args_process [err 0] <java.sql.SQLException> Error opening connection
[err 1] <db> connection open failed: info=gplx_key=sqlite;data source=/home/pi/xowa/user/xowa.user.anonymous.sqlite3;version=3;
[err 2] <db> db.engine:exec failed: url=data source=/home/pi/xowa/user/xowa.user.anonymous.sqlite3;version=3; sql=PRAGMA page_size = 4096;
[trace]:
  org.sqlite.SQLiteConnection.open(SQLiteConnection.java:179)
  org.sqlite.SQLiteConnection.<init>(SQLiteConnection.java:105)
  org.sqlite.JDBC.createConnection(JDBC.java:113)
  org.sqlite.JDBC.connect(JDBC.java:87)
  java.sql.DriverManager.getConnection(DriverManager.java:664)
  java.sql.DriverManager.getConnection(DriverManager.java:247)
  gplx.dbs.engines.Db_engine_sql_base.Conn_make_by_url(Unknown Source)
  gplx.dbs.engines.sqlite.Sqlite_engine.Conn_new(Unknown Source)
  gplx.dbs.engines.Db_engine_sql_base.New_stmt_exec(Unknown Source)
  gplx.dbs.engines.Db_engine_sql_base.Exec_as_int(Unknown Source)
  gplx.dbs.engines.Db_engine_sql_base.Exec_as_obj(Unknown Source)
  gplx.dbs.Db_conn.Exec_qry(Unknown Source)
  gplx.dbs.engines.sqlite.Sqlite_engine_.Pragma_page_size(Unknown Source)
  gplx.dbs.Db_conn_bldr_wkr__sqlite.New(Unknown Source)
  gplx.dbs.Db_conn_bldr.New(Unknown Source)
  gplx.dbs.Db_conn_bldr.Get_or_new(Unknown Source)
  gplx.xowa.users.data.Xou_db_mgr.Init_by_app(Unknown Source)
  gplx.xowa.users.Xoue_user.Init_by_app(Unknown Source)
  gplx.xowa.Xoae_app.Init_by_app(Unknown Source)
  gplx.xowa.Xoa_app_boot_mgr.Run_app(Unknown Source)
  gplx.xowa.Xoa_app_boot_mgr.Run(Unknown Source)
  gplx.xowa.Xoa_app_.Run(Unknown Source)
  gplx.xowa.Xowa_main.main(Unknown Source)
```
The stack is shown twice and xowa doesn't launch neither in http server of course.
I checked user right, no root user files, xowa is in a home directory as mention in the howTo.
I'm using the oracle jdk8u60 by default but i got the same result with openJDK.
I tried several time, using different approach, what did I miss?
Thank you

Edit: forgot to mention that i have the same result with version 1.9.3

**Comment by gnosygnu** [comment]
Thanks for the details!

Unfortunately, XOWA doesn't run on raspberry pI. XOWA uses xerial's sqlite-jdbc 3.7.15 which doesn't support pi. 

I see that there it may be supported in 3.8.11 but I don't think it's working. See: https://github.com/xerial/sqlite-jdbc/issues/49 . It looks like the developer needs help (unfortunately I don't have a pi device)

Once xerial looks like they support it, I'll update XOWA's copy.

**Comment by yonisolo** [comment]
Hi,
That what i though so i tried to recompiled with the updated jdbc driver and swt for arm but the result is the same, i included SWT because of this stack:
```
OS: linux
Java: 1.8.0_60 (32 bit)
Java path: /usr/lib/jvm/java-8-oracle/jre
XOWA: 2.9.4.1
XOWA path: /home/pi/xowaBuild/xowa_arm.jar

Error: java.lang.NoClassDefFoundError: org/eclipse/swt/graphics/Device
Stack: gplx.gfui.Gfui_kit_.Swt(Unknown Source)
gplx.xowa.gui.Xoa_gui_mgr.Invk(Unknown Source)
gplx.GfsCore_.Exec(Unknown Source)
gplx.GfsCore_.Exec(Unknown Source)
gplx.GfsCore_.Exec(Unknown Source)
gplx.GfsCore.ExecOne_to(Unknown Source)
gplx.xowa.apps.gfss.Xoa_gfs_mgr.Run_str_for(Unknown Source)
gplx.xowa.apps.gfss.Xoa_gfs_mgr.Run_str_for(Unknown Source)
gplx.xowa.apps.gfss.Xoa_gfs_mgr.Run_url_for(Unknown Source)
gplx.xowa.apps.gfss.Xoa_gfs_mgr.Run_url(Unknown Source)
gplx.xowa.Xoa_app_boot_mgr.Run_app(Unknown Source)
gplx.xowa.Xoa_app_boot_mgr.Run(Unknown Source)
gplx.xowa.Xoa_app_.Run(Unknown Source)
gplx.xowa.Xowa_main.main(Unknown Source)
```
Nothing significantly change, i will inform you if any progress.
I used this source as tip for the moment:
http://www.lucid-control.com/create-raspberry-pi-swt-gui-applications-with-java-and-eclipse/
and of course pick the sqlite-jdbc-3.8.11.1 drivers from  xerial repo (https://github.com/xerial/sqlite-jdbc)

**Comment by yonisolo** [comment]
I compiled the raspbian sqlite library and submitted it to @xerial for xerial/sqlite-jdbc#49

**Comment by gnosygnu** [comment]
Cool. Thanks for taking the initiative and helping out with sqlite-jdbc.

Yeah, you'll need SWT to compile XOWA b/c SWT is referenced directly from the XOWA code. I'm hoping that the SWT requirement was fine for ARM (i.e: no issues compiling). If it becomes problematic, I can look at separating XOWA into two libraries: a server-side one and a GUI one.

I'll see how things go in sqlite-jdbc and upgrade once it's been resolved.

**Comment by yonisolo** [comment]
no problem, since i want use it, it's quite normal to help community, there was no problem with SWT after installing the system requirement. However, in an achitectural way, it could be quite a good organisation if you have time and motivation to do it... Thanks for the follow up

**Comment by xerial** [comment]
I have just released sqlite-jdbc-3.8.11.2 with the @yonisolo's fix. Try this version.

**Comment by gnosygnu** [comment]
@xerial Thanks for the extremely quick patch! I downloaded it and will be testing it with over the next week. If all goes well, I'll add it to XOWA v2.10.2.

@yonisolo Let me know how the new jar works on Raspberry Pi. I'll pick up a Raspberry Pi over the next few weeks and work on adding a Rasperry Pi package (since it looks like I'll need a Raspberry Pi version of swt-gtk-3.8.0.jar)

**Comment by yonisolo** [comment]
Hi Please find in attachement the 2 jar for raspi SWT, found in the
/usr/share/java of my pi created while installing system library, hope
these will work. If I can help do not hesitate to ask me :)
I added a pass with 7z to bypass attachment security  "swt"

2015-10-04 3:09 GMT+02:00 gnosygnu <notifications@github.com>:

**Comment by gnosygnu** [comment]
Thanks! However, I didn't receive any attachment at either github or gmail. Maybe it got scrubbed because of the jar in the 7z? Do you want to try uploading them to another server, like perhaps https://filetea.me/default/ ?

Also, does that mean you were able to get XOWA running on a Raspberry PI? If so, how was it in terms of performance?

**Comment by yonisolo** [comment]
i sent them via gmail directly instead of the github no-reply, i didn't take the time to recompile xowa yet (another work in parallel), i'll try and come to you as soon as possible. I did test xowa with oracle mission control on debian, will try on the raspi2 to compare...

**Comment by gnosygnu** [comment]
Cool. Thanks. I received them via gmail. Sadly, filetea.me was inaccessible

I'll wait for your results before generating another package. Hope to get Raspberry Pi in next few weeks.

Thanks again for the jars.

**Comment by yonisolo** [comment]
It doesn't want to compie properly or I'm doing something wrong, I modified the sqlite jdbc in
```
xowa/140_dbs/lib/sqlite-jdbc-3.8.11.2.jar
```
replace the swt by mine in 
```
xowa/150_gfui/lib/swt.jar
```
and modified the ant build intructions:
modified the full name of the sqlite jar in the "compile_dbs" and "create_jar" targets
and launch the build with 
```
ant -buildfile build.xml -Dplat_name=arm -q
```
the build give some warning about _ identifiers but the jar is well created.
```
(use of '_' as an identifier might not be supported in releases after Java SE 8)
```
when I launch the created jar I got a classNotFound exception for sqlite
```
app init failed: args_process [err 0] <java.lang.ClassNotFoundException> org.sqlite.JDBC
[err 1] <db> could not load sqlite jdbc driver
[err 2] <db> db.engine:exec failed: url=data source=/home/pi/xowaBuild2/user/xowa.user.anonymous.sqlite3;version=3; sql=PRAGMA page_size = 4096;
[trace]:
  java.net.URLClassLoader.findClass(URLClassLoader.java:381)
  java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
  java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  java.lang.Class.forName0(Native Method)
  java.lang.Class.forName(Class.java:264)
  gplx.dbs.engines.sqlite.Sqlite_engine.Conn_new(Unknown Source)
  gplx.dbs.engines.Db_engine_sql_base.New_stmt_exec(Unknown Source)
  gplx.dbs.engines.Db_engine_sql_base.Exec_as_int(Unknown Source)
  gplx.dbs.engines.Db_engine_sql_base.Exec_as_obj(Unknown Source)
  gplx.dbs.Db_conn.Exec_qry(Unknown Source)
  gplx.dbs.engines.sqlite.Sqlite_engine_.Pragma_page_size(Unknown Source)
  gplx.dbs.Db_conn_bldr_wkr__sqlite.New(Unknown Source)
  gplx.dbs.Db_conn_bldr.New(Unknown Source)
  gplx.dbs.Db_conn_bldr.Get_or_new(Unknown Source)
  gplx.xowa.users.data.Xou_db_mgr.Init_by_app(Unknown Source)
  gplx.xowa.users.Xoue_user.Init_by_app(Unknown Source)
  gplx.xowa.Xoae_app.Init_by_app(Unknown Source)
  gplx.xowa.Xoa_app_boot_mgr.Run_app(Unknown Source)
  gplx.xowa.Xoa_app_boot_mgr.Run(Unknown Source)
  gplx.xowa.Xoa_app_.Run(Unknown Source)
  gplx.xowa.Xowa_main.main(Unknown Source)
```
Did I miss something? Thanks

**Comment by gnosygnu** [comment]
Thanks for the details. Responding to certain points:

> the build give some warning about _ identifiers but the jar is well created.

Yeah, I used a "_" as an identifier. I changed it for the upcoming 2.10.2 release, but it shouldn't affect your build

> when I launch the created jar I got a classNotFound exception for sqlite

Hmm... This means that it couldn't find the SQLite JDBC class. Maybe there was a linking error that got lost in all the "_" warnings?

I'll make a build later on today and see if there's anything with the build process.

Thanks.

**Comment by yonisolo** [comment]
Will redirect my output to a file and watch carrefully

**Comment by gnosygnu** [comment]
Ok. I think I was able to reproduce the problem. Basically, the sqlite-jdbc-3.8.11.2.jar needs to be copied to two locations:

* /xowa/dev/140_dbs/lib/sqlite-jdbc-3.8.11.2.jar
* /xowa/bin/any/java/jdbc/sqlite/sqlite-jdbc-3.8.11.2.jar

I think you got the 1st (/dev/) but not the 2nd (/bin/). If I don't copy the file to the 2nd location (/bin/) then I get the same error as you.

Something similar also needs to be done for the SWT jars.

* /xowa/dev/150_gfui/lib/swt.jar
* /xowa/bin/arm/swt/swt.jar

I'll change the build script to remove these duplicate references and only use the 2nd (/bin/)

Finally, just to be clear, I'd use the following steps:
* Create a new folder called /home/pi/xowaBuild3/
* Extract v2.9.4 to this folder. You'll have a folder like "/home/pi/xowaBuild3/bin/any" as well as many others
* Extract the source to this folder. You'll have a folder like "/home/pi/xowaBuild3/dev/400_xowa"  as well as many others
* Copy the sqlite jars here:
    * /home/pi/xowaBuild3/dev/140_dbs/lib/sqlite-jdbc-3.8.11.2.jar
    * /home/pi/xowaBuild3/bin/any/java/jdbc/sqlite/sqlite-jdbc-3.8.11.2.jar
* Copy the swt jars here:
   * /home/pi/xowaBuild3/dev/150_gfui/lib/swt.jar
   * /home/pi/xowaBuild3/bin/arm/swt/swt.jar
* Run the ant file

Hope this helps.

**Comment by yonisolo** [comment]
This helped a lot, the jdbc stage is ok but I got a new Unsatisfied link error while getting the connection, here is the trace:
```
Error: java.lang.UnsatisfiedLinkError: org.sqlite.core.NativeDB._open(Ljava/lang/String;I)V
Stack: org.sqlite.core.NativeDB._open(Native Method)
org.sqlite.core.DB.open(DB.java:161)
org.sqlite.core.CoreConnection.open(CoreConnection.java:220)
org.sqlite.core.CoreConnection.<init>(CoreConnection.java:76)
org.sqlite.jdbc3.JDBC3Connection.<init>(JDBC3Connection.java:24)
org.sqlite.jdbc4.JDBC4Connection.<init>(JDBC4Connection.java:23)
org.sqlite.SQLiteConnection.<init>(SQLiteConnection.java:45)
org.sqlite.JDBC.createConnection(JDBC.java:114)
org.sqlite.JDBC.connect(JDBC.java:88)
java.sql.DriverManager.getConnection(DriverManager.java:664)
java.sql.DriverManager.getConnection(DriverManager.java:247)
gplx.dbs.engines.Db_engine_sql_base.Conn_make_by_url(Unknown Source)
gplx.dbs.engines.sqlite.Sqlite_engine.Conn_new(Unknown Source)
gplx.dbs.engines.Db_engine_sql_base.New_stmt_exec(Unknown Source)
gplx.dbs.engines.Db_engine_sql_base.Exec_as_int(Unknown Source)
gplx.dbs.engines.Db_engine_sql_base.Exec_as_obj(Unknown Source)
gplx.dbs.Db_conn.Exec_qry(Unknown Source)
gplx.dbs.engines.sqlite.Sqlite_engine_.Pragma_page_size(Unknown Source)
gplx.dbs.Db_conn_bldr_wkr__sqlite.New(Unknown Source)
gplx.dbs.Db_conn_bldr.New(Unknown Source)
gplx.dbs.Db_conn_bldr.Get_or_new(Unknown Source)
gplx.xowa.users.data.Xou_db_mgr.Init_by_app(Unknown Source)
gplx.xowa.users.Xoue_user.Init_by_app(Unknown Source)
gplx.xowa.Xoae_app.Init_by_app(Unknown Source)
gplx.xowa.Xoa_app_boot_mgr.Run_app(Unknown Source)
gplx.xowa.Xoa_app_boot_mgr.Run(Unknown Source)
gplx.xowa.Xoa_app_.Run(Unknown Source)
gplx.xowa.Xowa_main.main(Unknown Source)
```
I tried with plat_name=ARM in uppercase too with the same result.
Hope we can do it!

**Comment by gnosygnu** [comment]
Hmm... It looks like this is a sqlite/jdbc issue. [The line of code](https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/core/DB.java#L161) appears to fail when calling the native ARM binary.

Maybe open a new issue with sqlite/jdbc and see if they have any advice?

Thanks.

**Comment by yonisolo** [comment]
Ok, @DavidCGG compile a snapshot which seems to work see xerial/sqlite-jdbc#58, after replacing yours by this one, the http_server is launch but the graphical interface need libswt-xulrunner-gtk which doesn't seems to be ported on arm yet, the libswt-cairo-gtk is not installed by default but is availlable, I'll send you the DavidCGG' snapshot by mail. You can close this issue for me.

**Comment by gnosygnu** [comment]
Thanks for the update, and glad to hear it worked!

Once you send me the copy of sqlite-jdbc-3.8.11.3-SNAPSHOT.jar, I'll build a basic ARM package. Hopefully it will be enough to work in HTTP Server mode (despite the non-working SWT parts)

**Comment by gnosygnu** [comment]
@yonisolo Again, many thanks for building the arm version of sqlite-jdbc. XOWA will now include ARM packages by default going forward. Let me know if there's anything else.

**Comment by yonisolo** [comment]
Thank you, what i did is not as great as your work!

**Comment by gnosygnu** [comment]
Respectfully disagree! Let's just say we both did pretty good. ;)

**Comment by syndycat** [comment]
Hi,

I have a Rasp3 and I downlaoded the latest sqlite-jdbc-3.8.11.2.jar. I have a similar issue when I open the connection: 

java.lang.UnsatisfiedLinkError: org.sqlite.core.NativeDB._open(Ljava/lang/String;I)V
        at org.sqlite.core.NativeDB._open(Native Method)
        at org.sqlite.core.DB.open(DB.java:161)

Can you share with me the fix for this issue?

Thank you,

**Comment by gnosygnu** [comment]
Hi! The latest sqlite-jdbc-3.8.11.2 jar at xerial's releases doesn't have the fix. We're still waiting on them to merge the pull request.

XOWA uses a modified version of sqlite-jdbc-3.8.11.2.jar that has yonisolo's fix. You can pull this version from the sqlite-jdbc jar in https://github.com/gnosygnu/xowa/releases/download/v3.4.2.1/xowa_app_arm_v3.4.2.1.zip . Specifically, it's in /xowa/bin/any/java/jdbc/sqlite/sqlite-jdbc-3.8.11.2.jar.

Let me know if you use this particular jar and it still fails. It shouldn't: both yonisolo and I tested with an earlier version (IIRC, https://github.com/gnosygnu/xowa/releases/download/v3.3.4.1/xowa_app_arm_v3.3.4.1.zip) and the jar hasn't changed since then.

Thanks!

**Comment by yonisolo** [comment]
Hi! it has been tested on banana pi and raspberry pi 2 which are ARMv7 . Raspberry 3 use armv8 Cortex A53 instructions,  i don't know if the cortex A53 is hard float or soft float, try the xowa fix. If it doesn't work, you must recompile the drivers with the option -march=armv8-a+crc instead of native with GCC which is specific to the ARM v8 architecture. see this manual page: https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html

**Comment by yonisolo** [comment]
@syndycat as @gnosygnu said:

> XOWA uses a modified version of sqlite-jdbc-3.8.11.2.jar that has yonisolo's fix. You can pull this   version from the sqlite-jdbc jar in https://github.com/gnosygnu/xowa/releases/download/v3.4.2.1/xowa_app_arm_v3.4.2.1.zip . Specifically, it's in /xowa/bin/any/java/jdbc/sqlite/sqlite-jdbc-3.8.11.2.jar.

unzip the archive and get the modified sqlite jdbc drivers /xowa/bin/any/java/jdbc/sqlite/sqlite-jdbc-3.8.11.2.jar.
If it works on your rpi 3, please let us know, thanks

**Comment by syndycat** [comment]
Hi everybody,

Thank you very much for the fast answers and jar file. You are the masters .... IT IS WORKING ON PI 3. !  
Now I am able to connect to database and do queries etc. 

However, I noticed another issue regarding INSERT INTO sql cmd that seems to have problems. For instance I have the following (all fields are type Text)

INSERT INTO contacts (msisdnid,lastTime,devicesStatus,devicesIP,devices,devicesLastTime) VALUES ('9999123456789',456166603,'2','','500259','456166603')

raises

 java.sql.SQLException: not implemented by SQLite JDBC driver
        at org.sqlite.jdbc3.JDBC3Statement.unused(JDBC3Statement.java:397)
        at org.sqlite.jdbc3.JDBC3Statement.executeUpdate(JDBC3Statement.java:408)
        at com.imalog.iOlaCore.Engine.SqliteDatabase.insert(SqliteDatabase.java:218)

If I copy & paste the above sql cmd into sqlite3 command line... the cmd  is working and the table has that new row. I know that this topic is not related wiht this thread, but do you have any idea why it is not working with this jdbc library ? 

Thank you,
Cata

**Comment by yonisolo** [comment]
try to use only
> INSERT INTO contacts VALUES ('9999123456789',456166603,'2','','500259','456166603')
if it works that's that the other syntax is not yet supported by this drivers

**Comment by gnosygnu** [comment]
@yonisolo: Thanks for the follow-up! I didn't know Raspberry PI 3 had different CPU instructions.

@syndycat: Yonisolo is correct. This looks like unsupported bevavior. As a guess, are you doing something like "conn.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);"? The stack frame of "org.sqlite.jdbc3.JDBC3Statement.executeUpdate(JDBC3Statement.java:408)" leads to https://github.com/xerial/sqlite-jdbc/blob/master/src/main/java/org/sqlite/jdbc3/JDBC3Statement.java#L407.

If you are doing this, then you can't: it is unsupported behavior. You may want to open an issue at xerial to get more info: https://github.com/xerial/sqlite-jdbc/issues

Hope this helps.

**Comment by yonisolo** [comment]
related https://github.com/xerial/sqlite-jdbc/issues/58
@gnosygnu @xerial, I cross referenced this issue not to repeat ourselves.

**Comment by syndycat** [comment]
Thank you guys.  Yes, that Statement.RETURN_GENERATED_KEYS connection cfg.  was the root cause. However,  Insert multiple values it is working fine too .... and regarding the generated keys, you can can get it from the result. 
Thank you for your statement around this (extra issue).

**Comment by gnosygnu** [comment]
@yonisolo Thanks!

@syndycat Thanks for the follow-up! Glad to help. I'm going to mark this ticket closed. Good luck
